add gazebo sawyer

hejia-zhang opened the pull request:
# Add sawyer simulation support and add gazebo environment

Design the ros environment support for rllab. Add sawyer simulation support and gazebo environment support.

Refer to: #49

hejia-zhang committed:
add gazebo sawyer

eric-heiden commented on contrib/gazebo/gazebo_env.py:
```diff
@@ -0,0 +1,156 @@
+"""
+This is gazebo base environment in rllab
```
Change to "Gazebo base environment in rllab"

eric-heiden commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,419 @@
+"""
+This is gazebo environment for a specific robot - sawyer.
```
Change to "Gazebo environment for the Sawyer robot"

eric-heiden commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,419 @@
+                              model_pose=Pose(position=Point(initial_qpos['table0'][0],
+                                                             initial_qpos['table0'][1],
+                                                             initial_qpos['table0'][2])),
+                              model_path=rospkg.RosPack().get_path('sawyer_sim_examples') + "/models/" + "cafe_table/model.sdf")
```
Here, and all other path combinations: use `os.path.join([dir1, dir2, ..., file])`

eric-heiden commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,419 @@
+    def _sample_goal(self):
+        """
+        Sample goals
+        :return:
```
Remove or complete this line

eric-heiden commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,419 @@
+                                                                                   self.target_range, size=3)
+            goal[2] = self.height_offset
+            if self.target_in_the_air and self.np_random_generator.uniform() < 0.5:
+                goal[2] += self.np_random_generator.uniform(0, 0.45)
```
You should write a comment here why at 50% of the cases you want to offset the goal's z position.

eric-heiden commented on contrib/gazebo/sawyer/pick_and_place.py:
```diff
@@ -0,0 +1,23 @@
+"""
+This is pick and place task environment for sawyer robot in rllab
```
"Pick-and-place task for the Sawyer robot"

eric-heiden commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,419 @@
+        else:
+            return -d
+
+    # Impletation for rllab env functions
```
*Implementation

eric-heiden commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,419 @@
+            GazeboEnv.delete_gazebo_model(model_name="block")
+            GazeboEnv.delete_gazebo_model(model_name="target")
+
+            # Kill gzclient, gzserver and roscore
```
Is there a case where we need the following deactivated code? If so, you should explain when to use it or else remove it.

eric-heiden commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,419 @@
+        """
+        Input
+        -----
+        control_mode: have 4 different control modes: position, velocity, trajectory-position, torque
```
There is no `control_mode` argument. Update the documentation with the current arguments and their meanings.

eric-heiden commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,419 @@
+        """
+
+        # Unpause simulation to make observation
+        # rospy.wait_for_service('/gazebo/unpause_physics')
```
Do we need this code?

eric-heiden reviewed (changes requested):
As you can see from the Travis build failure, there are some code formatting issues which you need to fix by running yapf on all your code.
From my understanding, `ik.py` and `sawyer_pnp.py` are demos from Rethink Robotics. So I would propose to move them to a separate folder like `contrib/gazebo/vendor_demos/`.
There are a few places where code blocks are commented out. It would be good to either remove them or comment why we need them. If they are needed, they should be made callable in separate functions.

hejia-zhang committed:
reorganize vendor demos code

hejia-zhang committed:
vendor demos

hejia-zhang committed:
reformatting

hejia-zhang committed:
1. Add push task; 2. add three control modes; 3. remove the gazebo start script which is commented out before; 4. add some comments

hejia-zhang committed:
reformatting

hejia-zhang committed:
reformatting

ryanjulian commented on contrib/gazebo/bug_reproducer.py:
```diff
@@ -0,0 +1,36 @@
+import rospy
+import intera_interface
```
this is not relevant to the feature and should be removed

ryanjulian commented on contrib/gazebo/gazebo_env.py:
```diff
@@ -0,0 +1,162 @@
+"""
```
Please put this file in the package contrib.gazebo.envs, to stay consistent with the other trees

ryanjulian commented on contrib/gazebo/gazebo_env.py:
```diff
@@ -0,0 +1,162 @@
+"""
+
+from rllab.envs.base import Env
+from rllab.spaces import Box
```
please use PEP8 import ordering

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+from .gazebo_env import GazeboEnv
+from rllab.spaces import Box
+import numpy as np
+import rospy
```
please use PEP8 import order

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+
+from intera_core_msgs.msg import JointLimits
+
+from gazebo_msgs.msg import (
```
why the multiline import style??

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+from geometry_msgs.msg import (Pose, Point)
+
+
+class ControlModeUnknownErr(Exception):
```
ControlModelUnknownError()

ryanjulian replied:
probably sufficient to use a NotImplementedError

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+        Input
+        -----
+        has_object: if there is object in this experiment
+        distance_threshold: this is used for sparse reward computation
```
please be more specific

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+        -----
+        has_object: if there is object in this experiment
+        distance_threshold: this is used for sparse reward computation
+        initial_qpos: dict, joint's positions and object's positions
```
joint positions and object positions

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+        has_object: if there is object in this experiment
+        distance_threshold: this is used for sparse reward computation
+        initial_qpos: dict, joint's positions and object's positions
+        reward_type: 'sparse' of 'dense'
```
rename `sparse_reward` and make it a bool

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+        distance_threshold: this is used for sparse reward computation
+        initial_qpos: dict, joint's positions and object's positions
+        reward_type: 'sparse' of 'dense'
+        target_in_the_ari: set to see if the target can be in the air
```
this is misspelled. also, what is the purpose of this param? do i ever want a target in the air?

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+        initial_qpos: dict, joint's positions and object's positions
+        reward_type: 'sparse' of 'dense'
+        target_in_the_ari: set to see if the target can be in the air
+        limb: the name of the limb
```
does this change? sawyer only has one limb.

hejia-zhang replied:
Although the sawyer only has one limb, we still need to pass in the limb's name(e.g. 'right') to intera interface. I think they are trying to make the package "intera interface" compatible with both of sawyer and baxter. I can hard-code the name inside the function. But I am not sure if it is good.

ryanjulian replied:
Nothing else in this change suggests it supports Baxter (and that would require a lot more work), so I think the scope is Sawyer-only. We can just hardcode.

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+        reward_type: 'sparse' of 'dense'
+        target_in_the_ari: set to see if the target can be in the air
+        limb: the name of the limb
+        end_point: the name of the end_point
```
how does this ever change?

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+        target_in_the_ari: set to see if the target can be in the air
+        limb: the name of the limb
+        end_point: the name of the end_point
+        target_range: used to sample the new target
```
i don't understand what this means.

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+        limb: the name of the limb
+        end_point: the name of the end_point
+        target_range: used to sample the new target
+        obj_range: used to sample the new position for object
```
i don't understand what this means

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+        print("Enabling robot...")
+        self._rs.enable()
+
+        def clean_shutdown():
```
why `clean_shutdown`? would i ever want an unclean shutdown?

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+    def update_model_states(self, data):
+        self.model_states = data
+
+    def compute_reward(self, achieved_goal, goal, info=None):
```
`reward()` is sufficient

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+
+    def compute_reward(self, achieved_goal, goal, info=None):
+        # Compute distance between goal and the achieved goal.
+        d = goal_distance(achieved_goal, goal)
```
what is the difference between the goal and the achieved goal?

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+        # Compute distance between goal and the achieved goal.
+        d = goal_distance(achieved_goal, goal)
+        if self.reward_type == 'sparse':
+            return -(d > self.distance_threshold).astype(np.float32)
```
you probably want a completion bonus to make training this more stable.

hejia-zhang replied:
Yeah! And I think for sparse reward completion bonus could be 1. But I am not sure how to set the completion bonus for dense reward.

ryanjulian replied:
@eric-heiden and I found between 10 and 100 to be sufficient for TRPO, IIRC.

You can make the completion bonus for sparse rewards the same. As the rewards are sparse, the relative magnitude is (mostly) irrelevant, but it is convenient to make it the same as the dense reward completion bonus, so that that magnitude of other types of rewards (e.g. entropy) is approximately the same between them.

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+"""
+
+from rllab.envs.base import Step
+from .gazebo_env import GazeboEnv
```
please do not use relative package imports

ryanjulian commented on contrib/gazebo/sawyer/pick_and_place.py:
```diff
@@ -0,0 +1,33 @@
+class SawyerPickAndPlaceEnv(gazebo_sawyer_env.GazeboSawyerEnv, Serializable):
+    def __init__(self, reward_type='sparse'):
+        Serializable.quick_init(self, locals())
+        initial_qpos = {
```
please break this out into a package-level constant INITIAL_QPOS rather than dropping it into the constructor.

hejia-zhang replied:
Great point! Is it ok if I define these constants in __init__.py? If not, could you give me some good examples?

ryanjulian replied:
Do not do anything noteworthy in `__init__.py`

Do it in package scope.

e.g.

`contrib/gazebo/sawyer/pick_and_place.py`:
```python
import stuff

INITIAL_QPOS = {
    'stuff': 0.0,
    'more_stuff': 1.0,
}

class SawyerPickAndPlaceEnv():
    # also some stuff
   
    def __init__():
        # more stuff here
```

ryanjulian commented on contrib/gazebo/sawyer/push.py:
```diff
@@ -0,0 +1,34 @@
+    def __init__(self, reward_type='dense'):
+        Serializable.quick_init(self, locals())
+        initial_qpos = {
+            'right_j0': -0.041662954890248294,
```
break out into INITIAL_QPOS

ryanjulian commented on contrib/gazebo/vendor_demos/ik.py:
```diff
@@ -0,0 +1,291 @@
+#!/usr/bin/env python
```
please do not copy code from other projects into the repo.

ryanjulian replied:
i.e. this file should be deleted.

ryanjulian commented on contrib/gazebo/gazebo_env.py:
```diff
@@ -0,0 +1,162 @@
+# Every environment class in rllab should inherit Serializable and call Serializable.quick_init(slef, locals())
+from rllab.core.serializable import Serializable
+
+from gazebo_msgs.srv import (SpawnModel, DeleteModel)
```
Please use PEP8 import style. One import per line, no ().

ryanjulian commented on contrib/gazebo/gazebo_env.py:
```diff
@@ -0,0 +1,162 @@
+
+    #-------------------------------------------------------
+    # Need to be implemented in specific robot env
+    def _sample_goal(self):
```
underscores denote a function which is private (thus should not be implemented by a subclass). you should remove the underscores.

ryanjulian commented on contrib/gazebo/gazebo_env.py:
```diff
@@ -0,0 +1,162 @@
+        """
+        raise NotImplementedError
+
+    def _env_setup(self, initial_qpos):
```
does this setup Gazebo or the rllab environment?

hejia-zhang replied:
Basically it does the setup work, like load the models and move the robot arm to the start position and get the initial gripper position, joint positions, object positions.

ryanjulian replied:
So Gazebo, then.

Does it need to be called during reset?

hejia-zhang replied:
No, just once in constructor of GazeboEnv

ryanjulian replied:
Okay, so should it be called _gazebo_setup()?

ryanjulian commented on contrib/gazebo/gazebo_env.py:
```diff
@@ -0,0 +1,162 @@
+    @property
+    def goal(self):
+        """
+        Returns _goal
```
You may omit docstrings for functions with obvious behavior. Also, please don't refer to private variables in docstrings. Imagine I am reading your docstrings but all of the code has been deleted--they still need to make sense.

ryanjulian commented on contrib/gazebo/gazebo_env.py:
```diff
@@ -0,0 +1,162 @@
+
+    # gazebo env needed functions
+    #---------------------------------------------------------------
+    def seed(self, seed=None):
```
please use the global seed provided by rllab. it allows us to control the seed for the entire system at once, making results more consistent. i think rllab should seed numpy for you, anyway.

ryanjulian commented on contrib/gazebo/gazebo_env.py:
```diff
@@ -0,0 +1,162 @@
+        # Load target SDF
+        model_xml = ''
+        with open(model_path, "r") as model_file:
+            model_xml = model_file.read().replace('\n', '')
```
why the replace? does the XML parser really choke on newlines?

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+            lower_bounds = np.array(self._joint_limits.position_lower[3:10])
+            upper_bounds = np.array(self._joint_limits.position_upper[3:10])
+        elif self._control_mode == 'velocity':
+            lower_bounds = np.zeros(7)
```
np.zeros_like(self._joint_limits.position_lower[3:10])

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+
+        self._action_space = Box(
+            np.concatenate((lower_bounds, np.array([0]))),
+            np.concatenate((upper_bounds, np.array([100]))))
```
if these are chosen arbitrarily, please pull them out into constants at the top of the file.

hejia-zhang replied:
Do you mean the lower/upper bounds? They are chosen according to the robot's joint limits. The last element of lower/upper bounds is the percentage that the gripper is open.

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+
+        # set joint angles
+        joint_angle_cmds = {
+            'right_j0': action[0],
```
```python
joint_angle_cmds = {}
for i in range(7):
    joint_angle_cmds['right_j{}'.format(i)'] = action[i]
```

also, if it is _always_ right_j[n], how is there a parameter for which limb to use?

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+        if self._control_mode != "position":
+            rospy.wait_for_service('/gazebo/pause_physics')
+            try:
+                pause = rospy.ServiceProxy('/gazebo/pause_physics',
```
please predefine all your ServiceProxys either on import or construction, and make them class or package members

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+            if self._control_mode != 'position':
+                rospy.wait_for_service('/gazebo/unpause_physics')
+                try:
+                    unpause = rospy.ServiceProxy('/gazebo/unpause_physics',
```
predefine

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+            if self._control_mode != 'position':
+                rospy.wait_for_service('/gazebo/pause_physics')
+                try:
+                    pause = rospy.ServiceProxy('/gazebo/pause_physics',
```
predefine

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+        self.target_range = target_range
+        self.target_in_the_air = target_in_the_air
+        self.obj_range = obj_range
+        self.start_angles = {
```
uhh isn't this equivalent to `self.start_angles = initial_qpos`?

hejia-zhang replied:
initial_qpos also includes the object's positions

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+        self._observation_space = Box(
+            -np.inf, np.inf, shape=self._get_obs()['observation'].shape)
+
+    def _set_action_space(self):
```
can't this just be the definition of action_space()?

hejia-zhang replied:
What do you mean?

ryanjulian replied:
Why not just put this code in action_space() rather than storing the result in `self._action_space`?

hejia-zhang replied:
I think if action_space() was called frequently, storing the result would save some time. But I am not sure if action_space() would be called frequently.

ryanjulian replied:
http://wiki.c2.com/?PrematureOptimization

hejia-zhang replied:
Great!

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+        print("Getting robot state...")
+        self._rs = intera_interface.RobotEnable(intera_interface.CHECK_VERSION)
+        self._init_state = self._rs.state().enabled
+        print("Enabling robot...")
```
please don't change hardware state in constructors, or do any other long-running RPC operations. you can do this in reset() instead.

ryanjulian commented on contrib/gazebo/gazebo_env.py:
```diff
@@ -0,0 +1,162 @@
+        # Spawn Target URDF
+        rospy.wait_for_service('/gazebo/spawn_sdf_model')
+        try:
+            spawn_sdf = rospy.ServiceProxy('/gazebo/spawn_sdf_model',
```
predefine

ryanjulian commented on contrib/gazebo/gazebo_env.py:
```diff
@@ -0,0 +1,162 @@
+        # Spawn Target URDF
+        rospy.wait_for_service('/gazebo/spawn_urdf_model')
+        try:
+            spawn_urdf = rospy.ServiceProxy('/gazebo/spawn_urdf_model',
```
predefine

ryanjulian commented on contrib/gazebo/gazebo_env.py:
```diff
@@ -0,0 +1,162 @@
+    def delete_gazebo_model(cls, model_name):
+        rospy.wait_for_service('/gazebo/delete_model')
+        try:
+            delete_model = rospy.ServiceProxy('/gazebo/delete_model',
```
predefine

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+    return np.linalg.norm(goal_a - goal_b, axis=-1)
+
+
+def rate_limited(max_per_second: int):
```
please move this into some utility package (e.g. contrib.gazebo.util)

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+            "/gazebo/model_states", ModelStates, timeout=5)
+        rospy.Subscriber('/gazebo/model_states', ModelStates,
+                         self.update_model_states)
+        GazeboEnv.__init__(self, initial_qpos=initial_qpos)
```
the ordering of a lot of this doesn't make sense. please call the superclass either directly at the top or bottom. setup related things together, and separate them logically.

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+        if self._control_mode != 'position':
+            rospy.wait_for_service('/gazebo/unpause_physics')
+            try:
+                unpause = rospy.ServiceProxy('/gazebo/unpause_physics',
```
predefine

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+            # Unpause simulation to make observation
+            rospy.wait_for_service('/gazebo/unpause_physics')
+            try:
+                unpause = rospy.ServiceProxy('/gazebo/unpause_physics',
```
predefine

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+            new_pose=Pose(
+                position=Point(x=self.goal[0], y=self.goal[1], z=self.goal[
+                    2])),
+            reference_frame='world')
```
i thought there's a parameter for the reference frame?

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+        """
+
+        if self._control_mode != 'position':
+            # Unpause simulation to make observation
```
hmm why do we pause the simulation again? the real robot won't pause.

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+                position=Point(initial_qpos['table0'][0], initial_qpos[
+                    'table0'][1], initial_qpos['table0'][2])),
+            model_path=osp.join(
+                rospkg.RosPack().get_path('sawyer_sim_examples'), "models",
```
please use only " " strings or ' ' strings throughout a file. we prefer ' ' for everything but large blocks of text to be read by humans.

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+        rospy.sleep(1.0)
+
+    def _get_obs(self):
+        print("Getting observation...")
```
Please remove spurious print statements.

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+        gripper_lvel = np.array(
+            self._limb.endpoint_velocity()['linear']).copy()
+        gripper_avel = np.array(
+            self._limb.endpoint_velocity()['angular']).copy()
```
are these copies really necessary?

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+
+        # gazebo data message
+        if self.has_object:
+            object_pos = np.zeros(0)
```
equivalent to `np.array()`

ryanjulian commented on contrib/gazebo/gazebo_sawyer_env.py:
```diff
@@ -0,0 +1,499 @@
+                        object_pose.orientation.z, object_pose.orientation.w
+                    ]).copy()
+                except ValueError as err:
+                    pass
```
uhhhh why does a ValueError arise an why would i ignore it?

ryanjulian commented on contrib/gazebo/sawyer/pick_and_place.py:
```diff
@@ -0,0 +1,33 @@
+"""
+
+from contrib.gazebo import gazebo_sawyer_env
+from rllab.core.serializable import Serializable
```
PEP8 import order

hejia-zhang replied:
I went through the imports section in PEP8 and I found I should group imports into three groups(standard packages, third party packages, local packages/sources). Do you mean the two imports here belong to different groups. I thought they are both local packages. Let me know if I was wrong!

ryanjulian replied:
We are going to extend that logic, since rllab has its own hierarchy of import paths.

For example (obviously comments would not appear in real code):
```python
# python standard library
import os

# third-party packages
import tensorflow as tf
import numpy as np

# rllab top-level tree
from rllab.envs.mujoco.swimmer_env import SwimmerEnv

# rllab sandboxes
from sandbox.rocky.tf.policies.gaussian_mlp_policy import GaussianMLPPolicy

# project-local code (e.g. contrib in this case)
from contrib.gazebo import gazebo_sawyer_env
```

ryanjulian commented on contrib/gazebo/sawyer/push.py:
```diff
@@ -0,0 +1,34 @@
+Push task for the sawyer robot
+"""
+
+from contrib.gazebo import gazebo_sawyer_env
```
PEP8 import order

ryanjulian reviewed (changes requested):
This is almost there but needs some polish.

Some overall comments
* Please provide a README.md which provides all of the commands needed to setup a completely clean system (which already has a working rllab) to be ready to run this code. That includes ROS, intera, and whatever else is needed. If you have to add pip packages to the conda environment, include that too. Everything.

* The README.md should also tell me exactly which commands to run to start Gazebo (or the real robot ROS) before I start rllab.

* Where used, please normalize the docstrings to conform to PEP287

* Please avoid calling long-running operations (e.g. RPC calls) in constructors if at all possible. It may be impossible to avoid sometimes.

* NEVER enable a real robot in a constructor. That needs to be an intentional act.

* GazeboSawyerEnv is very lengthy, and I suspect it could be split into 2 classes. Perhaps you could make a GazeboSawyer class which abstracts the ROS details and use it inside GazeboSawyerEnv

* contrib.gazebo and Gazebo* are not quite the right names for this package, since it's perfectly capable of controlling the real robot. perhaps it should be contrib.ros

* There is no need to prefix the name of every class in this package with Gazebo--that's what the package name is for.

* Consider how you would redesign this PR if you needed it to support both a Gazebo-backed Sawyer and a real Sawyer using the same environment interface, and the only thing which changes is a constructor parameter to SawyerPickAndPlaceEnv(). This will be one of your next PRs, anyway. We can merge this code now and you can refactor it later, but I definitely won't be accepting a PR for real Sawyer support which copies tons of this code rather than reusing almost all of it (sans some parameters).

ryanjulian commented:
Related to #7 and #8

ryanjulian commented on contrib/gazebo/sawyer/pick_and_place.py:
```diff
@@ -0,0 +1,33 @@
+from rllab.core.serializable import Serializable
+
+
+class SawyerPickAndPlaceEnv(gazebo_sawyer_env.GazeboSawyerEnv, Serializable):
```
With few exceptions, files should contain a single class, and the file name should match the class name.

ryanjulian commented on contrib/gazebo/sawyer/push.py:
```diff
@@ -0,0 +1,34 @@
+"""
+Push task for the sawyer robot
```
Please also put this file (and all other environments) into contrib.gazebo.envs, so as the match the top-level package organization of the rest of rllab.

hejia-zhang committed:
modified code

hejia-zhang committed:
modified

hejia-zhang committed:
modified

hejia-zhang committed:
modified

hejia-zhang committed:
modify

ryanjulian commented on contrib/ros/envs/ros_env.py:
```diff
@@ -0,0 +1,52 @@
+import numpy as np
+
+# Every environment class in rllab should inherit Serializable and call Serializable.quick_init(self, locals())
```
remove this comment

ryanjulian commented on contrib/ros/envs/sawyer_env.py:
```diff
@@ -0,0 +1,356 @@
+import numpy as np
+import rospy
+import rospkg
+from gazebo_msgs.msg import ModelStates
```
nit: alphabetical order

hejia-zhang replied:
Do you mean the order of rospy and rospkg? I am not sure about how to sort import ... statement and from ... import ... statement.

ryanjulian replied:
consider `from ... import ...` and `import ...` to be the same for alphabetical purposes, and sort alphabetically based on the first appearance of a package name.

e.g. this block would be sorted
```python
from gazebo_msgs.msg import ModelStates
import intera_interface
import numpy as np
import rospkg
import rospy
```
because `gazebo_msgs.msg` comes before `intera_interface` alphabetically regardless of the `from`, and because `rospkg` comes before `rospy`

ryanjulian commented on contrib/ros/envs/sawyer_env.py:
```diff
@@ -0,0 +1,356 @@
+from contrib.ros.util.common import rate_limited
+from contrib.ros.util.gazebo import Gazebo
+
+import intera_interface
```
this is a 3rd party dep, so it should be grouped along with numpy and ros deps above.

ryanjulian commented on contrib/ros/envs/sawyer_env.py:
```diff
@@ -0,0 +1,356 @@
+import intera_interface
+
+
+def goal_distance(goal_a, goal_b):
```
please make this underscore-private `_goal_distance()`

ryanjulian commented on contrib/ros/envs/sawyer_env.py:
```diff
@@ -0,0 +1,356 @@
+            initial_robot_joint_pos, control_mode=robot_control_mode)
+        if not self._robot.enabled:
+            logger.log("Please enable the sawyer!")
+            raise Exception
```
1. Don't raise anonymous exceptions--it should have some string description of what went wrong.
1. I think you should use `RuntimeError` for this.
1. Please move the long-running ROS/robot setup code to a function `initialize()` and make it part of the `RosEnv` interface.

ryanjulian commented on contrib/ros/envs/sawyer_env.py:
```diff
@@ -0,0 +1,356 @@
+            logger.log("Robot has been enabled!")
+
+        self._real = real
+        if not self._real:
```
Please make the default state real, so this would change to

```python
self._simulated = simulated
if self._simulated:
   # ...
```

ryanjulian commented on contrib/ros/envs/sawyer_env.py:
```diff
@@ -0,0 +1,356 @@
+        else:
+            # rospy.Subscriber('/gazebo/model_states', ModelStates,
+            #                  self.update_model_states)
+            # TODO
```
Please file all TODOs at GitHub issues and include the issue number in your TODO.

For example, let's say I have a TODO which I file on GitHub, and GitHub assigns it issue number 1234.

```python
# TODO(gh/1234): Fix some stuff
```

ryanjulian commented on contrib/ros/envs/sawyer_env.py:
```diff
@@ -0,0 +1,356 @@
+        self.target_in_the_air = target_in_the_air
+        self.obj_range = obj_range
+
+        def clean_shutdown():
```
Please define this as a private function of the class, rather than as a nested function. Call it `shutdown()`.

ryanjulian replied:
`shutdown()` might also be part of the `RosEnv` interface.

ryanjulian commented on contrib/ros/envs/sawyer_env.py:
```diff
@@ -0,0 +1,356 @@
+        :return: the copy of new sampled goal
+        """
+        if self.has_object:
+            goal = self.initial_gripper_pos[:
```
this is very weird indentation. you can fix it manually and use `# yapf: disable`

https://github.com/google/yapf has more info

ryanjulian commented on contrib/ros/robots/sawyer.py:
```diff
@@ -0,0 +1,123 @@
+import intera_interface
+import numpy as np
+import rospy
+from intera_core_msgs.msg import JointLimits
```
Please order alphabetically

hejia-zhang replied:
Not sure how to sort from ... statement and import ... statement

ryanjulian replied:
See comment above. Ignore the `from` and sort only based on package names (and then by class/function names, if applicable)

ryanjulian commented on contrib/ros/README.md:
```diff
@@ -0,0 +1,138 @@
+# rllab.contrib.ros Instructions
+## Introduction
+Most of the current environments for reinfocement learning training(e.g. gym's environments) are not particularly 
```
missing space

jonashen replied:
reinfocement -> reinforcement

ryanjulian commented on contrib/ros/README.md:
```diff
@@ -0,0 +1,138 @@
+# rllab.contrib.ros Instructions
+## Introduction
+Most of the current environments for reinfocement learning training(e.g. gym's environments) are not particularly 
+designed for robot learning. Consequently, most of their environments are not for continuous action space task. And 
```
don't start a sentence with And.

hejia-zhang replied:
Gotcha!

ryanjulian commented on contrib/ros/README.md:
```diff
@@ -0,0 +1,138 @@
+##### Initialize rosdep
+rosdep enables you to easily install system dependencies for source you want to compile and is required to run some core components in ROS.
+    
+    $ sudo rosdep init
```
You can make shell code blocks to force this all to display nicely. 

https://guides.github.com/features/mastering-markdown/

<pre lang="no-highlight"><code>
 
```sh
ls -la
cd ~
fortune
sudo apt update && sudo apt upgrade
```
</code></pre>

becomes
```sh
ls -la
cd ~
fortune
sudo apt update && sudo apt upgrade
```

hejia-zhang replied:
Good point!

ryanjulian commented on contrib/ros/README.md:
```diff
@@ -0,0 +1,138 @@
+    $ sudo apt-get install python-rosinstall
+#### CREATE DEVELOPMENT WORKSPACE
+##### Create ROS Workspace
+    $ mkdir -p ~/<YOUR_FAVORITE_WORKSPACE_NAME>/src
```
You can make it easier on users by having them define <YOUR_FAVORITE_WORKSPACE_NAME> as a bash variable instead, then using it throughout the tutorial.

For example
```sh
export YOUR_ROS_WS=~/path/to/your/ros/workspace/src
```

Then, everywhere else in the tutorial use the variable, e.g.
```sh
mkdir -p ~/$YOUR_ROS_WS/src
```

hejia-zhang replied:
Got it!

ryanjulian commented on contrib/ros/README.md:
```diff
@@ -0,0 +1,138 @@
+    $ source /opt/ros/kinetic/setup.bash
+##### Build
+    $ cd ~/<YOUR_FAVORITE_WORKSPACE_NAME>
+    $ catkin_make -DPYTHON_EXECUTABLE:FILEPATH=<path_to_your_rllab_env_python_interp>
```
Can you provide more info to the user on how to locate `<path_to_your_rllab_env_python_interp>`? They are likely not experts on conda. You can probably extract it into a variable to help them.

```sh
cd $YOUR_ROS_WS
export RLLAB_PYTHON=`conda env rllab3 --something`
catkin_make -DPYTHON_EXECUTABLE:FILEPATH=$RLLAB_PYTHON
```

jonashen replied:
Ex.  $RLLAB_PYTHON = "/home/sawyer/anaconda2/envs/rllab3/bin/python"

hejia-zhang replied:
What I used:
```bash
$ source activate rllab3
$ export RLLAB_PYTHON=`which python`
$ source deactivate
```

ryanjulian commented on contrib/ros/README.md:
```diff
@@ -0,0 +1,138 @@
+Remove sawyer_simulator, if you don't need it:
+
+    packages=(intera_sdk)
+##### Execute script
```
Please be very specific about where to ``cd`` into before executing the script, and provide a code example.

jonashen replied:
./sawyer_2to3.sh

_If permission denied, run:_
sudo chmod +x sawyer_2to3.sh

hejia-zhang replied:
What I added:
```bash
$ rllab_path=<the_path_to_rllab> ./$rllab_path/contrib/ros/scripts/sawyer_2to3.sh
```

ryanjulian reviewed (changes requested):
This looks great, very professional. I added some comments with small change requests.

Since it is such a large process, I want someone else to try your setup tutorial to provide feedback.

@jonashen can you please follow Hejia's setup tutorial on a clean machine to make sure it works? You can use Zhanpeng's machine. Please also provide a code review here with suggested clarifications and spelling/grammar corrections.

jonashen commented on contrib/ros/README.md:
```diff
@@ -0,0 +1,138 @@
+it is also hard for robot learning researchers to use the same environment interface when they want to test their 
+algorithms on real robots. 
+
+rllab.contrib.ros's environments take a different approach. It's targeting at provide the same interfaces for both of the 
```
It's targeted towards providing the same interface for training both simulated and real-life robots.

jonashen commented on contrib/ros/README.md:
```diff
@@ -0,0 +1,138 @@
+    $ echo "source /opt/ros/kinetic/setup.bash" >> ~/.bashrc
+##### Build
+    $ cd ~/<YOUR_FAVORITE_WORKSPACE_NAME>
+    $ catkin_make
```
Base path: /home/sawyer/ros_ws_test
Source space: /home/sawyer/ros_ws_test/src
Build space: /home/sawyer/ros_ws_test/build
Devel space: /home/sawyer/ros_ws_test/devel
Install space: /home/sawyer/ros_ws_test/install
Creating symlink "/home/sawyer/ros_ws_test/src/CMakeLists.txt" pointing to "/opt/ros/kinetic/share/catkin/cmake/toplevel.cmake"
####
#### Running command: "cmake /home/sawyer/ros_ws_test/src -DCATKIN_DEVEL_PREFIX=/home/sawyer/ros_ws_test/devel -DCMAKE_INSTALL_PREFIX=/home/sawyer/ros_ws_test/install -G Unix Makefiles" in "/home/sawyer/ros_ws_test/build"
####
-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Using CATKIN_DEVEL_PREFIX: /home/sawyer/ros_ws_test/devel
-- Using CMAKE_PREFIX_PATH: /opt/ros/kinetic
-- This workspace overlays: /opt/ros/kinetic
-- Found PythonInterp: /home/sawyer/anaconda2/bin/python (found version "2.7.14") 
-- Using PYTHON_EXECUTABLE: /home/sawyer/anaconda2/bin/python
-- Using Debian Python package layout
-- Using empy: /usr/bin/empy
-- Using CATKIN_ENABLE_TESTING: ON
-- Call enable_testing()
-- Using CATKIN_TEST_RESULTS_DIR: /home/sawyer/ros_ws_test/build/test_results
-- Found gmock sources under '/usr/src/gmock': gmock will be built
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found gtest sources under '/usr/src/gmock': gtests will be built
-- Using Python nosetests: /usr/bin/nosetests-2.7
ImportError: "from catkin_pkg.package import parse_package" failed: No module named catkin_pkg.package
Make sure that you have installed "catkin_pkg", it is up to date and on the PYTHONPATH.
CMake Error at /opt/ros/kinetic/share/catkin/cmake/safe_execute_process.cmake:11 (message):
  execute_process(/home/sawyer/anaconda2/bin/python
  "/opt/ros/kinetic/share/catkin/cmake/parse_package_xml.py"
  "/opt/ros/kinetic/share/catkin/cmake/../package.xml"
  "/home/sawyer/ros_ws_test/build/catkin/catkin_generated/version/package.cmake")
  returned error code 1
Call Stack (most recent call first):
  /opt/ros/kinetic/share/catkin/cmake/catkin_package_xml.cmake:74 (safe_execute_process)
  /opt/ros/kinetic/share/catkin/cmake/all.cmake:151 (_catkin_package_xml)
  /opt/ros/kinetic/share/catkin/cmake/catkinConfig.cmake:20 (include)
  CMakeLists.txt:52 (find_package)


-- Configuring incomplete, errors occurred!
See also "/home/sawyer/ros_ws_test/build/CMakeFiles/CMakeOutput.log".
See also "/home/sawyer/ros_ws_test/build/CMakeFiles/CMakeError.log".
Invoking "cmake" failed

jonashen replied:
catkin_make -DPYTHON_EXECUTABLE:FILEPATH=/home/sawyer/anaconda2/envs/rllab3/bin/python

jonashen commented on contrib/ros/README.md:
```diff
@@ -0,0 +1,138 @@
+##### Save and Close intera.sh script
+##### Initialize your SDK environment
+    $ cd ~/<YOUR_FAVORITE_WORKSPACE_NAME>
+    $ ./intera.sh
```
Only executes for real robots, need to make a note of it.

hejia-zhang replied:
Deleted it.

jonashen commented on contrib/ros/README.md:
```diff
@@ -0,0 +1,138 @@
+### Start the gazebo simulation
+- This section is only for simulation user
+#### Copy the sawyer simulation launch file and simulation world file
+    $ cp <rllab.contrib.ros.envs.sawyer.sawyer_learning.launch, rllab.contrib.ros.envs.sawyer.sawyer_world_learning.launch> <YOUR_FAVORITE_WORKSPACE_PATH>/src/sawyer_simulator/sawyer_gazebo/launch
```
cp <YOUR_PATH_TO_RLLAB>/rllab/contrib/ros/envs/sawyer/{sawyer_learning.launch,sawyer_world_learning.launch} <YOUR_FAVORITE_WORKSPACE_PATH>/src/sawyer_simulator/sawyer_gazebo/launch

jonashen commented on contrib/ros/README.md:
```diff
@@ -0,0 +1,138 @@
+    $ ./intera.sh sim
+    $ export PYTHONPATH=/opt/ros/kinetic/lib/python2.7/dist-packages:<YOUR_FAVORITE_WORKSPACE_PATH>/devel/lib/python3/dist-packages
+#### Launch the gazebo sawyer
+    $ roslaunch sawyer_gazebo sawyer_learning.launch
```
apt-install python-rospkg

jonashen replied:
pip install rospkg

jonashen replied:
pip install catkin_pkg

jonashen replied:
_No module named defusedxml.xmlrpc_
pip install defusedxml

jonashen replied:
Make sure to tell the users launching GAZEBO for the first time that the program will take a long time to open, but it's normal because it's downloading extra packages. Also depends on the computer running it

jonashen replied:
@ryanjulian Where should @hjzh4 list the dependencies? At the end of the README?

ryanjulian replied:
Inline as apt/pip installation instructions, plus a brief bullet list of major deps at the top section where he mentions the system configuration this assumes.

hejia-zhang replied:
Listed the python packages dependencies like
## Support Platform
| Robot        | Workstation OS | ROS    | Python | Simulation Platform | 
|:------------:|:--------------:| :-----:|:------:|:-------------------:|
| Sawyer       | Ubuntu16.04    | Kinetic| 3      |Gazebo 7.0           |
### Major python packages dependencies
- catkin_pkg
- rospkg
- defusedxml
Added the installation intructions
##### Install Python Dependecies
```bash
$ source activate rllab3
$ pip install catkin_pkg
$ pip install rospkg
$ pip install defusedxml
```

jonashen commented on contrib/ros/README.md:
```diff
@@ -0,0 +1,138 @@
+    $ cp <rllab.contrib.ros.envs.sawyer.sawyer_learning.world> <YOUR_FAVORITE_WORKSPACE_PATH>/src/sawyer_simulator/sawyer_gazebo/worlds/
+#### Set the environment variables
+    $ cd ~/<YOUR_FAVORITE_WORKSPACE_NAME>
+    $ ./intera.sh sim
```
SHELL=/bin/bash ./intera.sh sim

hejia-zhang replied:
Added.

jonashen commented on contrib/ros/README.md:
```diff
@@ -0,0 +1,138 @@
+- This section is only for simulation user
+#### Copy the sawyer simulation launch file and simulation world file
+    $ cp <rllab.contrib.ros.envs.sawyer.sawyer_learning.launch, rllab.contrib.ros.envs.sawyer.sawyer_world_learning.launch> <YOUR_FAVORITE_WORKSPACE_PATH>/src/sawyer_simulator/sawyer_gazebo/launch
+    $ cp <rllab.contrib.ros.envs.sawyer.sawyer_learning.world> <YOUR_FAVORITE_WORKSPACE_PATH>/src/sawyer_simulator/sawyer_gazebo/worlds/
```
cp <YOUR_PATH_TO_RLLAB>/rllab/contrib/ros/envs/sawyer/sawyer_learning.world <YOUR_FAVORITE_WORKSPACE_PATH>/src/sawyer_simulator/sawyer_gazebo/worlds

ryanjulian replied:
please define bash variables for `$RLLAB` and `$WORKSPACE` instead, as mentioned above.

hejia-zhang replied:
What I added:
## Define bash variables which will be used throughout the tutorial
### Define the path of python interpreter used in rllab conda env as bash variable
```bash
$ source activate rllab3
$ export RLLAB_PYTHON=`which python`
$ source deactivate
```
### Define your favorite workspace name as bash variable 
    $ export YOUR_ROS_WS=/path/to/your/ros/workspace/
### Define your rllab path as bash variable
    $ export RLLAB=/path/to/your/rllab/

jonashen commented on contrib/ros/envs/example_launchers/trpo_gazebo_sawyer_pnp.py:
```diff
@@ -0,0 +1,39 @@
+from rllab.baselines.linear_feature_baseline import LinearFeatureBaseline
```
_No module named theano.tensor.nnet_

jonashen commented on contrib/ros/envs/example_launchers/trpo_gazebo_sawyer_push.py:
```diff
@@ -0,0 +1,39 @@
+from rllab.baselines.linear_feature_baseline import LinearFeatureBaseline
```
No module named theano.tensor.nnet

jonashen commented on contrib/ros/README.md:
```diff
@@ -0,0 +1,138 @@
+### Run the training script
+#### Set the environment variables
+    $ cd ~/<YOUR_FAVORITE_WORKSPACE_NAME>
+    $ ./intera.sh sim
```
SHELL=/bin/bash ./intera.sh sim

hejia-zhang replied:
Added.

jonashen commented on contrib/ros/README.md:
```diff
@@ -0,0 +1,138 @@
+    $ ./intera.sh sim
+    $ source activate rllab3
+#### Run the training script
+    $ python <rllab.contrib.ros.envs.example_launchers>/<launcher>
```
Ex. python <YOUR_PATH_TO_RLLAB>/rllab/contrib/ros/envs/example_launchers/trpo_gazebo_sawyer_pnp.py

jonashen replied:
Fix _No module named 'contrib.ros'_

hejia-zhang replied:
What I added:
#### Run the training script
**Make sure that $RLLAB is in your $PYTHONPATH** \
Ex.

    $ python $RLLAB/contrib/ros/envs/example_launchers/trpo_gazebo_sawyer_pnp.py

hejia-zhang committed:
modified

hejia-zhang committed:
modified

hejia-zhang committed:
change YOUR_ROS_WS to ROS_ws

hejia-zhang committed:
modified

hejia-zhang committed:
Add pip dependecy opencv-python

hejia-zhang committed:
Add something

ryanjulian commented on contrib/ros/README.md:
````diff
@@ -0,0 +1,194 @@
+$ export RLLAB_PYTHON=`which python`
+$ source deactivate
+```
+### Define your favorite workspace name as bash variable 
````
merge these three sub headings (defining variables) into one set of commands in a shell block.

ryanjulian commented on contrib/ros/envs/sawyer_env.py:
```diff
@@ -0,0 +1,346 @@
+        if self._simulated:
+            self.model_states = data
+        else:
+            # TODO(gh/2: Finish the 'update_model_states' function for real robot)
```
```python
# TODO(gh/???): Finish the 'update_model_states' function for real robot)
```
Where `???` is the issue number github assigns your todo when you file it here:https://github.com/ryanjulian/rllab/issues

hejia-zhang replied:
Got it I am using gh/8 Sawyer runtime support

ryanjulian commented on contrib/ros/envs/sawyer_env.py:
```diff
@@ -0,0 +1,346 @@
+                             self.update_model_states)
+            self._gazebosrv = Gazebo()
+        else:
+            # TODO(gh/1: Add vicon system utility for models' states update)
```
```python
# TODO(gh/???): Add vicon system utility for models' states update)
```
Where `???` is the issue number github assigns your todo when you file it here:https://github.com/ryanjulian/rllab/issues

ryanjulian commented on contrib/ros/envs/sawyer_env.py:
```diff
@@ -0,0 +1,346 @@
+
+        # Randomize start position of object
+        if self.has_object:
+            object_xypos = self.initial_gripper_pos[:2]
```
please make this formatting less weird and disable/enable yapf in this region with ``# yapf: disable`` and ``# yapf: enable``

ryanjulian replied:
It seems that most of your formatting problem is huge expressions. You can make them smaller, and the code more clear, by using temporary variables. Then you won't have to fight the formatter.

ryanjulian commented on contrib/ros/envs/sawyer_env.py:
```diff
@@ -0,0 +1,346 @@
+            self._gazebosrv.load_gazebo_sdf_model(
+                model_name='cafe_table',
+                model_pose=Pose(
+                    position=Point(initial_model_pos['table0'][
```
please don't include huge expressions in kwargs. Make temporary variables for all these.

hejia-zhang replied:
Okay

ryanjulian commented on contrib/ros/envs/sawyer_env.py:
```diff
@@ -0,0 +1,346 @@
+                    rospkg.RosPack().get_path('sawyer_sim_examples'), 'models',
+                    'target/model.sdf'))
+        else:
+            # TODO(gh/4: Finish 'initial_setup' function for real robot)
```
```python
# TODO(gh/???): Finish 'initial_setup' function for real robot)
```
Where `???` is the issue number github assigns your todo when you file it here:https://github.com/ryanjulian/rllab/issues

ryanjulian commented on contrib/ros/envs/sawyer_env.py:
```diff
@@ -0,0 +1,346 @@
+            # TODO(gh/4: Finish 'initial_setup' function for real robot)
+            pass
+
+    # yapf: disable
```
Rather than using `# yapf: disable`, just define temporary variables for your 3 random variables.

ryanjulian commented on contrib/ros/envs/ros_env.py:
```diff
@@ -0,0 +1,57 @@
+    def __init__(self):
+        Serializable.quick_init(self, locals())
+
+        self.np_random_generator = np.random.RandomState(get_seed())
```
Why do you store a random generator? numpy's seed is globally-initialized by rllab.

ryanjulian commented on contrib/ros/envs/sawyer_env.py:
```diff
@@ -0,0 +1,346 @@
+            initial_robot_joint_pos, control_mode=robot_control_mode)
+        if not self._robot.enabled:
+            raise RuntimeError(
+                '[Error] SawyerEnv.__init__: The robot is not enabled!')
```
Remove the "[Error] SawyerEnv.__init__:'"stuff. (1) It will be incorrect if this code is ever refactored and (2) all of that information is provided by Python when a RuntimeError is thrown.

ryanjulian replied:
Does the robot need to be enabled for the constructor to finish? why?

ryanjulian commented on contrib/ros/envs/sawyer_env.py:
```diff
@@ -0,0 +1,346 @@
+
+        RosEnv.__init__(self)
+        # long-running ROS/robot setup
+        self.initialize()
```
I meant that `initialize()` should be called by the user _after_ the constructor is finished. If that is not possible in the current design, please make a TODO for it.

ryanjulian commented on contrib/ros/envs/sawyer_env.py:
```diff
@@ -0,0 +1,346 @@
+                reference_frame='world')
+            self._reset_sim()
+        else:
+            # TODO(gh/3: Finish the reset function for real robot)
```
```python
# TODO(gh/???): Finish the reset function for real robot
```
Where `???` is the issue number github assigns your todo when you file it here:https://github.com/ryanjulian/rllab/issues

ryanjulian commented on contrib/ros/envs/sawyer_env.py:
```diff
@@ -0,0 +1,346 @@
+                    object_pos = np.array([
+                        object_pose.position.x, object_pose.position.y,
+                        object_pose.position.z
+                    ]).copy()
```
why on earth do you need to copy an array you just constructed into a variable you own?

ryanjulian commented on contrib/ros/envs/sawyer_env.py:
```diff
@@ -0,0 +1,346 @@
+                    ]).copy()
+                except ValueError as err:
+                    logger.log(
+                        'Error _get_obs(484): The model named "block" does not exist!'
```
Yet another reason not to include line numbers and function names in errors: it is already wrong.

ryanjulian commented on contrib/ros/envs/sawyer_env.py:
```diff
@@ -0,0 +1,346 @@
+                        object_pose.orientation.x, object_pose.orientation.y,
+                        object_pose.orientation.z, object_pose.orientation.w
+                    ]).copy()
+                except ValueError as err:
```
What is the right thing to do here? If there is no model named block, how can we continue anyway?

ryanjulian commented on contrib/ros/envs/sawyer_env.py:
```diff
@@ -0,0 +1,346 @@
+        obs = np.concatenate((robot_obs, object_pos, object_ori))
+
+        return {
+            'observation': obs.copy(),
```
again, why do you need to copy()?

ryanjulian commented on contrib/ros/robots/sawyer.py:
```diff
@@ -0,0 +1,123 @@
+
+    @property
+    def enabled(self):
+        logger.log("If the robot has been enabled!")
```
this is a useless logger line. please remove.

ryanjulian commented on contrib/ros/envs/sawyer_env.py:
```diff
@@ -0,0 +1,346 @@
+        obs = self._get_obs()
+        initial_observation = obs['observation']
+
+        return initial_observation.copy()
```
why do you copy()?

ryanjulian commented on contrib/ros/robots/sawyer.py:
```diff
@@ -0,0 +1,123 @@
+        obs = np.concatenate(
+            (gripper_pos, gripper_ori, gripper_lvel, gripper_avel,
+             gripper_force, gripper_torque, robot_joint_angles))
+        return obs.copy()
```
why do you copy()?

ryanjulian commented on contrib/ros/robots/sawyer.py:
```diff
@@ -0,0 +1,123 @@
+            lower_bounds = np.zeros_like(self._joint_limits.effort[3:10])
+            upper_bounds = np.array(self._joint_limits.effort[3:10])
+        else:
+            # should raise an unknown mode exception
```
This is actually a ValueError

ryanjulian commented on contrib/ros/util/common.py:
```diff
@@ -0,0 +1,31 @@
+import threading
+import time
+from functools import wraps
```
alphabetical order

ryanjulian commented on contrib/ros/util/gazebo.py:
```diff
@@ -0,0 +1,65 @@
+import rospy
```
these should all be grouped together. also, alphabetical order.

ryanjulian commented on contrib/ros/envs/ros_env.py:
```diff
@@ -0,0 +1,57 @@
+import numpy as np
+import rospy
+
+# Every environment class in rllab should inherit Serializable and call Serializable.quick_init(self, locals())
```
remove this comment, it's useless.

ryanjulian reviewed (changes requested):
Please address of the comments here. If it takes more than 10 minutes to address any of them, file a GitHub issue for it instead and leave a TODO.

We will merge this ASAP and then you can work on all of the TODOs.

hejia-zhang committed:
modified

hejia-zhang committed:
format

hejia-zhang committed:
format

hejia-zhang committed:
Add sawyer simulation support and add gazebo environment

Design the ros environment support for rllab. Add sawyer simulation support and gazebo environment support.

Refer to: #49

hejia-zhang committed:
format

hejia-zhang commented on contrib/ros/README.md:
```diff
@@ -0,0 +1,199 @@
+    $ mkdir -p $ROS_WS/src
+##### Source ROS Setup
+    $ source /opt/ros/kinetic/setup.bash
+You can also add this to you .bashrc by using
```
your

hejia-zhang commented on contrib/ros/README.md:
````diff
@@ -0,0 +1,199 @@
+$ wstool merge sawyer_robot/sawyer_robot.rosinstall
+$ wstool update
+```
+##### Download the dependecies
````
dependencies

hejia-zhang committed:
fix spell

hejia-zhang committed:
Merge branch 'integration' into gazebo_sawyer
